bootstrap/kubeadm/api/v1alpha4/conversion.go

```diff
@@ -17,20 +17,42 @@ limitations under the License.
 package v1alpha4
 
 import (
+	apiconversion "k8s.io/apimachinery/pkg/conversion"
 	"sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1beta1"
+	utilconversion "sigs.k8s.io/cluster-api/util/conversion"
 	"sigs.k8s.io/controller-runtime/pkg/conversion"
 )
 
 func (src *KubeadmConfig) ConvertTo(dstRaw conversion.Hub) error {
 	dst := dstRaw.(*v1beta1.KubeadmConfig)
 
-	return Convert_v1alpha4_KubeadmConfig_To_v1beta1_KubeadmConfig(src, dst, nil)
+	if err := Convert_v1alpha4_KubeadmConfig_To_v1beta1_KubeadmConfig(src, dst, nil); err != nil {
+		return err
+	}
+
+	// Manually restore data.
+	restored := &v1beta1.KubeadmConfig{}
+	if ok, err := utilconversion.UnmarshalData(src, restored); err != nil || !ok {
+		return err
+	}
+
+	dst.Spec.Ignition = restored.Spec.Ignition
+
+	return nil
 }
 
 func (dst *KubeadmConfig) ConvertFrom(srcRaw conversion.Hub) error {
 	src := srcRaw.(*v1beta1.KubeadmConfig)
 
-	return Convert_v1beta1_KubeadmConfig_To_v1alpha4_KubeadmConfig(src, dst, nil)
+	if err := Convert_v1beta1_KubeadmConfig_To_v1alpha4_KubeadmConfig(src, dst, nil); err != nil {
+		return err
+	}
+
+	if err := utilconversion.MarshalData(src, dst); err != nil {
+		return err
```

Comment:
@vincepri I made all the changes as you have mentioned. One last thing that is stopping me is that the FuzzyConversion tests for v1alpha4 fails with the following error:

```go
ok      sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1alpha3       98.282s
--- FAIL: TestFuzzyConversion (69.37s)
    --- FAIL: TestFuzzyConversion/for_KubeadmControlPlane (32.22s)
        --- FAIL: TestFuzzyConversion/for_KubeadmControlPlane/spoke-hub-spoke (0.26s)
            conversion.go:208: 
                Expected success, but got an error:
                    <*errors.errorString | 0xc00035e270>: {
                        s: "error decoding string from json: invalid character 'Ç' in string escape code",
                    }                  
                    error decoding string from json: invalid character 'Ç' in string escape code
    --- FAIL: TestFuzzyConversion/for_KubeadmControlPlaneTemplate (37.15s)
        --- FAIL: TestFuzzyConversion/for_KubeadmControlPlaneTemplate/spoke-hub-spoke (2.03s)
            conversion.go:208:                                                                           
                Expected success, but got an error:                                                      
                    <*errors.errorString | 0xc0003f7780>: {
                        s: "error decoding string from json: unexpected end of JSON input",          
                    }                                                                                    
                    error decoding string from json: unexpected end of JSON input
FAIL
FAIL    sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1alpha4       69.468s
```

This error is coming from this line.

Reply:
Digging deeper I see that the code that errors out is [here](https://github.com/kubernetes-sigs/structured-merge-diff/blob/87c6f2e7efba6350467ed69b5fe2268c225e18b3/value/reflectcache.go#L219).

Here is a sample run:

```
"}忍ɄǂŕPEHl鉄Ǟď5ɻ"渉.N壞鑖ĿŻʋlŞƱ[Ĳ咋Ɩ捧ǟv"

src obj:

&v1beta1.KubeadmConfig{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"ƚBÌ戞n)©ʀ:褊鰊L牁2Ʃ$", GenerateName:"殌)óÀ蛠ʂĵĳy黣ĭ翸Ĺ嶧ƀÇǑ", Namespace:"+ƍƙɜȽ*黯BQ~庙ɸŸA疏Ɯ士Ȋ", SelfLink:"", UID:"ɤ灎吮ĩ>ȶ鳆7", ResourceVersion:"13134269316646846690", Generation:-3222315741621318315, CreationTimestamp:time.Date(2017, time.November, 13, 7, 21, 32, 0, time.Local), DeletionTimestamp:time.Date(2097, time.September, 4, 10, 5, 30, 0, time.Local), DeletionGracePeriodSeconds:(*int64)(0xc0001a5608), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:"C;ű Y%ɿ鴡ȥ;ɹ", ManagedFields:[]v1.ManagedFieldsEntry(nil)}, Spec:v1beta1.KubeadmConfigSpec{ClusterConfiguration:(*v1beta1.ClusterConfiguration)(0xc0000ef8c0), InitConfiguration:(*v1beta1.InitConfiguration)(0xc000296c60), JoinConfiguration:(*v1beta1.JoinConfiguration)(nil), Files:[]v1beta1.File{v1beta1.File{Path:"劁焕犟U樬靑ð礿 ʅǲ6漚岒Ò", Owner:"", Permissions:"鋍WɎ胬Ɓ~ʭ.@絔QÍ咝磒 Ƶ噅5", Encoding:"", Content:"ƠǞå乁QɐhKĠ*6Ƴy", ContentFrom:(*v1beta1.FileSource)(0xc000349580)}}, DiskSetup:(*v1beta1.DiskSetup)(nil), Mounts:[]v1beta1.MountPoints(nil), PreKubeadmCommands:[]string(nil), PostKubeadmCommands:[]string(nil), Users:[]v1beta1.User{}, NTP:(*v1beta1.NTP)(0xc0003495e0), Format:"", Verbosity:(*int32)(0xc0001a5bc0), UseExperimentalRetryJoin:true, Ignition:(*v1beta1.IgnitionSpec)(nil)}, Status:v1beta1.KubeadmConfigStatus{Ready:false, DataSecretName:(*string)(0xc00068e2a0), FailureReason:"ȏƛY飣", FailureMessage:"²鬰Ƨɣ樚幛ǋv岫旂ʛčtǽ口钦", ObservedGeneration:-1239634301236346895, Conditions:v1beta1.Conditions{}}}

dst obj:

&v1alpha4.KubeadmConfig{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"ƚBÌ戞n)©ʀ:褊鰊L牁2Ʃ$", GenerateName:"殌)óÀ蛠ʂĵĳy黣ĭ翸Ĺ嶧ƀÇǑ", Namespace:"+ƍƙɜȽ*黯BQ~庙ɸŸA疏Ɯ士Ȋ", SelfLink:"", UID:"ɤ灎吮ĩ>ȶ鳆7", ResourceVersion:"13134269316646846690", Generation:-3222315741621318315, CreationTimestamp:time.Date(2017, time.November, 13, 7, 21, 32, 0, time.Local), DeletionTimestamp:time.Date(2097, time.September, 4, 10, 5, 30, 0, time.Local), DeletionGracePeriodSeconds:(*int64)(0xc0001a5608), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:"C;ű Y%ɿ鴡ȥ;ɹ", ManagedFields:[]v1.ManagedFieldsEntry(nil)}, Spec:v1alpha4.KubeadmConfigSpec{ClusterConfiguration:(*v1alpha4.ClusterConfiguration)(0xc0000ef8c0), InitConfiguration:(*v1alpha4.InitConfiguration)(0xc000296c60), JoinConfiguration:(*v1alpha4.JoinConfiguration)(nil), Files:[]v1alpha4.File{v1alpha4.File{Path:"劁焕犟U樬靑ð礿 ʅǲ6漚岒Ò", Owner:"", Permissions:"鋍WɎ胬Ɓ~ʭ.@絔QÍ咝磒 Ƶ噅5", Encoding:"", Content:"ƠǞå乁QɐhKĠ*6Ƴy", ContentFrom:(*v1alpha4.FileSource)(0xc000349580)}}, DiskSetup:(*v1alpha4.DiskSetup)(nil), Mounts:[]v1alpha4.MountPoints(nil), PreKubeadmCommands:[]string(nil), PostKubeadmCommands:[]string(nil), Users:[]v1alpha4.User{}, NTP:(*v1alpha4.NTP)(0xc0003495e0), Format:"", Verbosity:(*int32)(0xc0001a5bc0), UseExperimentalRetryJoin:true}, Status:v1alpha4.KubeadmConfigStatus{Ready:false, DataSecretName:(*string)(0xc00068e2a0), FailureReason:"ȏƛY飣", FailureMessage:"²鬰Ƨɣ樚幛ǋv岫旂ʛčtǽ口钦", ObservedGeneration:-1239634301236346895, Conditions:v1alpha4.Conditions{}}}

err:
error decoding string from json: invalid character 'æ' after top-level value

```

The error says that the character `æ` is invalid but it does not exist in any of the strings being processed. I am confused what is going wrong here.

Reply:
So turns out this is happening because of bootstraptoken, I wonder why is this not being ignored like the rest.

```go
sv: &v1beta1.BootstrapTokenString{ID:"閰ʬ/EŬȺ苸o骴琝挻ª", Secret:"ʪʠ@M禢,躠\"?"}
dv: interface {}(nil)
here is the data:

"閰ʬ/EŬȺ苸o骴琝挻ª.ʪʠ@M禢,躠"?"
--------------- error is coming from here -----------------
src:
&v1beta1.KubeadmConfig{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"鼍侴¡h妁ƉĂʒ·", GenerateName:"-ɥ>蹧慌斵剚ɮ睭#呐ɹǭż", Namespace:"ɤW9", SelfLink:"âț騾壕ȖY撵", UID:"藴*枻L毕l", ResourceVersion:"4286917835756494355", Generation:4801911080450449014, CreationTimestamp:time.Date(2035, time.March, 5, 22, 4, 44, 0, time.Local), DeletionTimestamp:time.Date(1986, time.December, 12, 6, 20, 45, 0, time.Local), DeletionGracePeriodSeconds:(*int64)(0xc0003dc3f0), Labels:map[string]string{"ÔT疍騲醘ýť旈鄷¸!篊Ȟɱ":"Ðȶ§e碧酊榤Š饽Ų½oǭ迿\"qh齇"}, Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:"碞逭|籽Ė牧ʍƻŰŢƇ", ManagedFields:[]v1.ManagedFieldsEntry{}}, Spec:v1beta1.KubeadmConfigSpec{ClusterConfiguration:(*v1beta1.ClusterConfiguration)(0xc00029af20), InitConfiguration:(*v1beta1.InitConfiguration)(0xc00002f3f0), JoinConfiguration:(*v1beta1.JoinConfiguration)(nil), Files:[]v1beta1.File{}, DiskSetup:(*v1beta1.DiskSetup)(0xc00055ea50), Mounts:[]v1beta1.MountPoints{v1beta1.MountPoints{}}, PreKubeadmCommands:[]string{}, PostKubeadmCommands:[]string{"@=ʑ,¶ò"}, Users:[]v1beta1.User{v1beta1.User{Name:"{", Gecos:(*string)(0xc00007d7c0), Groups:(*string)(nil), HomeDir:(*string)(0xc00007d7e0), Inactive:(*bool)(0xc0003dc8a0), Shell:(*string)(nil), Passwd:(*string)(nil), PrimaryGroup:(*string)(0xc00007d800), LockPassword:(*bool)(0xc0003dc8d8), Sudo:(*string)(0xc00007d820), SSHAuthorizedKeys:[]string(nil)}}, NTP:(*v1beta1.NTP)(nil), Format:"h*őĳkÆȌ", Verbosity:(*int32)(nil), UseExperimentalRetryJoin:true, Ignition:(*v1beta1.IgnitionSpec)(nil)}, Status:v1beta1.KubeadmConfigStatus{Ready:true, DataSecretName:(*string)(0xc00007d850), FailureReason:"轐倷Ƶ榳轏Ǉ魼#Y黇撶晰mř梃Ʊ羡ƾ鲰", FailureMessage:"ǬʣSŎ:@儯A", ObservedGeneration:-8920981233074323261, Conditions:v1beta1.Conditions{v1beta1.Condition{Type:"<Ƈf-貘嬻EB", Status:"橯莁C¿氃~臋O脰r[ę忕", Severity:"KMÌ節#漿潊Ƣ霜冬ʜ[.r¬尔炎ʫ", LastTransitionTime:time.Date(1995, time.August, 3, 8, 14, 14, 0, time.Local), Reason:"z", Message:"\\s熌爱h_o"}}}}

dst:
&v1alpha4.KubeadmConfig{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"鼍侴¡h妁ƉĂʒ·", GenerateName:"-ɥ>蹧慌斵剚ɮ睭#呐ɹǭż", Namespace:"ɤW9", SelfLink:"âț騾壕ȖY撵", UID:"藴*枻L毕l", ResourceVersion:"4286917835756494355", Generation:4801911080450449014, CreationTimestamp:time.Date(2035, time.March, 5, 22, 4, 44, 0, time.Local), DeletionTimestamp:time.Date(1986, time.December, 12, 6, 20, 45, 0, time.Local), DeletionGracePeriodSeconds:(*int64)(0xc0003dc3f0), Labels:map[string]string{"ÔT疍騲醘ýť旈鄷¸!篊Ȟɱ":"Ðȶ§e碧酊榤Š饽Ų½oǭ迿\"qh齇"}, Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:"碞逭|籽Ė牧ʍƻŰŢƇ", ManagedFields:[]v1.ManagedFieldsEntry{}}, Spec:v1alpha4.KubeadmConfigSpec{ClusterConfiguration:(*v1alpha4.ClusterConfiguration)(0xc00029af20), InitConfiguration:(*v1alpha4.InitConfiguration)(0xc00002f3f0), JoinConfiguration:(*v1alpha4.JoinConfiguration)(nil), Files:[]v1alpha4.File{}, DiskSetup:(*v1alpha4.DiskSetup)(0xc00055ea50), Mounts:[]v1alpha4.MountPoints{v1alpha4.MountPoints{}}, PreKubeadmCommands:[]string{}, PostKubeadmCommands:[]string{"@=ʑ,¶ò"}, Users:[]v1alpha4.User{v1alpha4.User{Name:"{", Gecos:(*string)(0xc00007d7c0), Groups:(*string)(nil), HomeDir:(*string)(0xc00007d7e0), Inactive:(*bool)(0xc0003dc8a0), Shell:(*string)(nil), Passwd:(*string)(nil), PrimaryGroup:(*string)(0xc00007d800), LockPassword:(*bool)(0xc0003dc8d8), Sudo:(*string)(0xc00007d820), SSHAuthorizedKeys:[]string(nil)}}, NTP:(*v1alpha4.NTP)(nil), Format:"h*őĳkÆȌ", Verbosity:(*int32)(nil), UseExperimentalRetryJoin:true}, Status:v1alpha4.KubeadmConfigStatus{Ready:true, DataSecretName:(*string)(0xc00007d850), FailureReason:"轐倷Ƶ榳轏Ǉ魼#Y黇撶晰mř梃Ʊ羡ƾ鲰", FailureMessage:"ǬʣSŎ:@儯A", ObservedGeneration:-8920981233074323261, Conditions:v1alpha4.Conditions{v1alpha4.Condition{Type:"<Ƈf-貘嬻EB", Status:"橯莁C¿氃~臋O脰r[ę忕", Severity:"KMÌ節#漿潊Ƣ霜冬ʜ[.r¬尔炎ʫ", LastTransitionTime:time.Date(1995, time.August, 3, 8, 14, 14, 0, time.Local), Reason:"z", Message:"\\s熌爱h_o"}}}}

err:
error decoding string from json: invalid character '?' after top-level value
```